airflow/utils/db.py

```diff
@@ -702,9 +727,33 @@ def upgradedb():
         for err in errs:
             log.error("Automatic migration is not available\n%s", err)
         return
+
+    dialect = session.connection().dialect
+    if dialect.name == 'postgresql':
+        log.info('Acquiring lock on database for upgrade')
+        session.connection().execute('select PG_ADVISORY_LOCK(2);')
+
+    if dialect.name == 'mysql' and dialect.server_version_info >= (
+        5,
+        6,
+    ):
+        log.info('Acquiring lock on database for upgrade')
+        session.connection().execute("select GET_LOCK('db_upgrade',1800);")
+
     command.upgrade(config, 'heads')
     add_default_pool_if_not_exists()
 
+    if dialect.name == 'postgresql':
+        log.info('Releasing lock on database for upgrade')
+        session.connection().execute('select PG_ADVISORY_UNLOCK(2);')
+
+    if dialect.name == 'mysql' and dialect.server_version_info >= (
+        5,
+        6,
+    ):
+        log.info('Releasing lock on database for upgrade')
+        session.connection().execute("select RELEASE_LOCK('db_upgrade');")
```

Comment:
This being duplicated three times makes me think we should wrap the logic, maybe into a context manager.

Reply:
Sorry for not following up too quickly.
What do you mean by context manager?

Do you mean creating own context manager like:
@contextlib.contextmanager
def create_session():
    """Contextmanager that will create and teardown a session."""
    session = settings.Session()
    try:
        yield session
        session.commit()
    except Exception:
        session.rollback()
        raise
    finally:
        session.close()
?

Reply:
For `select RELEASE_LOCK(`db_upgrade`)` etc.

```python
@contextmanager
def lockdb(session, key):
    conn = session.connection()
    dialect = conn.dialect

    log.info('Acquiring lock on database for upgrade')
    if dialect.name == "postgresql":
        conn.execute(f'select PG_ADVISORY_LOCK({key});')
    elif dialect.name == "mysql" and dialect.server_version_info >= (5, 6):
        session.connection().execute("select GET_LOCK('db_upgrade', 1800);")

    yield

    if dialect.name == "postgresql":
        conn.execute(f'select PG_ADVISORY_UNLOCK({key});')
    elif dialect.name == "mysql" and dialect.server_version_info >= (5, 6):
        session.connection().execute("select GET_LOCK('db_upgrade', 1800);")
```

so we can do

```python
with lockdb(session, 1):
    upgradedb()
```

Some additional management (e.g. a globally incremental counter) to eliminate `key` can be useful as well.

Reply:
Also, would be good to use it as decorator like
```
@lock_db
@provide_session
def initdb(session=None):
```
And same for `upgradedb` method.

What do you think guys?

Reply:
The decorator usage could be a bit difficult since the function requires some arguments, but +1 if it’s not too complicated to make it work.

Reply:
I have sent some tested code.